metaflow/plugins/events_decorator.py

```diff
@@ -0,0 +1,298 @@
+import json
+import time
+import re
+
+from metaflow import current
+from metaflow.decorators import FlowDecorator
+from metaflow.exception import MetaflowException
+from metaflow.util import is_stringish
+
+# TODO: Support dynamic parameter mapping through a context object that exposes
+#       flow name and user name similar to parameter context
+
+
+class TriggerDecorator(FlowDecorator):
+    """
+    Specifies the event(s) that this flow depends on.
+
+    ```
+    @trigger(event='foo')
+    ```
+    or
+    ```
+    @trigger(events=['foo', 'bar'])
+    ```
+
+    Additionally, you can specify the parameter mappings
+    to map event payload to Metaflow parameters for the flow.
+    ```
+    @trigger(event={'name':'foo', 'parameters':{'my_param': 'event_field'})
+    ```
+    or
+    ```
+    @trigger(events=[{'name':'foo', 'parameters':{'my_param_1': 'event_field_1'},
+                     {'name':'bar', 'parameters':{'my_param_2': 'event_field_2'}])
+    ```
+
+    Parameters
+    ----------
+    event : str or dict, optional
+        Event dependency for this flow.
+    events : List[str or dict], optional
+        Events dependency for this flow.
+    options : dict, optional
+        Backend-specific configuration for tuning eventing behavior.
+    """
+
+    name = "trigger"
+    defaults = {
+        "event": None,
+        "events": [],
+        "options": {},
+    }
+
+    def flow_init(
+        self,
+        flow_name,
+        graph,
+        environment,
+        flow_datastore,
+        metadata,
+        logger,
+        echo,
+        options,
+    ):
+        self.triggers = []
+        if sum(map(bool, (self.attributes["event"], self.attributes["events"]))) > 1:
+            raise MetaflowException(
+                "Specify only one of *event* or *events* "
+                "attributes in *@trigger* decorator."
+            )
+        elif self.attributes["event"]:
+            # event attribute supports the following formats -
+            #     1. event='table.prod_db.members'
+            #     2. event={'name': 'table.prod_db.members',
+            #               'parameters': {'alpha': 'member_weight'}}
+            if is_stringish(self.attributes["event"]):
+                self.triggers.append({"name": str(self.attributes["event"])})
+            elif isinstance(self.attributes["event"], dict):
+                if "name" not in dict(self.attributes["event"]):
+                    raise MetaflowException(
+                        "The *event* attribute for *@trigger* is missing the "
+                        "*name* key."
+                    )
+                self.triggers.append(self.attributes["event"])
```

Comment:
we could check the other fields here too. The error message hints at the format needed but we don't actually check.

Reply:
fine. it's just inconsistent with the error message but not a big deal.